drivers/mspi/mspi_ambiq_ap4.c

```diff
@@ -0,0 +1,1505 @@
+		if (data->dev_id != dev_id) {
+			LOG_INST_ERR(cfg->log, "%u, config failed, must be the same device.",
+				     __LINE__);
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if ((param_mask & (~(MSPI_DEVICE_CONFIG_FREQUENCY | MSPI_DEVICE_CONFIG_IO_MODE |
+				     MSPI_DEVICE_CONFIG_CE_NUM | MSPI_DEVICE_CONFIG_DATA_RATE |
+				     MSPI_DEVICE_CONFIG_CMD_LEN | MSPI_DEVICE_CONFIG_ADDR_LEN)))) {
+			LOG_INST_ERR(cfg->log, "%u, config type not supported.", __LINE__);
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_FREQUENCY) {
+			hal_dev_cfg.eClockFreq = mspi_set_freq(cfg, dev_cfg->freq);
+			if (hal_dev_cfg.eClockFreq == 0) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_CLOCK_CONFIG,
+						  &hal_dev_cfg.eClockFreq);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, failed to configure eClockFreq.",
+					     __LINE__);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+			data->dev_cfg.freq = dev_cfg->freq;
+		}
+
+		if ((param_mask & MSPI_DEVICE_CONFIG_IO_MODE) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_CE_NUM) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_DATA_RATE)) {
+			hal_dev_cfg.eDeviceConfig = mspi_set_line(
+				cfg, dev_cfg->io_mode, dev_cfg->data_rate, dev_cfg->ce_num);
+			if (hal_dev_cfg.eDeviceConfig == AM_HAL_MSPI_FLASH_MAX) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_disable(data->mspiHandle);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, fail to disable MSPI, code:%d.",
+					     __LINE__, ret);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_DEVICE_CONFIG,
+						  &hal_dev_cfg.eDeviceConfig);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, failed to configure device.", __LINE__);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_enable(data->mspiHandle);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, fail to enable MSPI, code:%d.",
+					     __LINE__, ret);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+
+			data->dev_cfg.freq = dev_cfg->io_mode;
+			data->dev_cfg.data_rate = dev_cfg->data_rate;
+			data->dev_cfg.ce_num = dev_cfg->ce_num;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_CMD_LEN) {
+			if (dev_cfg->cmd_length > AM_HAL_MSPI_INSTR_2_BYTE + 1 ||
+			    dev_cfg->cmd_length == 0) {
+				LOG_INST_ERR(cfg->log, "%u, invalid cmd_length.", __LINE__);
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			hal_dev_cfg.eInstrCfg = dev_cfg->cmd_length - 1;
+			data->dev_cfg.cmd_length = dev_cfg->cmd_length;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_ADDR_LEN) {
+			if (dev_cfg->addr_length > AM_HAL_MSPI_ADDR_4_BYTE + 1 ||
+			    dev_cfg->addr_length == 0) {
+				LOG_INST_ERR(cfg->log, "%u, invalid addr_length.", __LINE__);
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			hal_dev_cfg.eAddrCfg = dev_cfg->addr_length - 1;
+			data->dev_cfg.addr_length = dev_cfg->addr_length;
+		}
+	} else {
```

Comment:
```suggestion
	} else {
		if (data->dev_id != dev_id) {
			ret = pinctrl_apply_state(cfg->pcfg,
						  PINCTRL_STATE_PRIV_START + dev_id->dev_idx);
			if (ret) {
				goto e_return;
			}
		}
```
You must check the dev_id.